templates/boltzmann/middleware/honeycomb.ts

```diff
@@ -9,77 +10,151 @@ import beeline from 'honeycomb-beeline'
 import isDev from 'are-we-dev'
 void `{% endif %}`;
 
+let otelTracer: OtelTracer | null = null;
+
+function isHoneycomb (env: typeof process.env): boolean {
+  return !!env.HONEYCOMB_WRITEKEY
+}
+
+function isOtel (env: typeof process.env): boolean {
+  if (isHoneycomb(env) && env.HONEYCOMB_API_HOST) {
+    return /^grpc:\/\//.test(env.HONEYCOMB_API_HOST)
+  }
+  return false
+}
+
+function getOtelTracer() {
+  if (!otelTracer) {
+    otelTracer = otelTrace.getTracer('boltzmann', '1.0.0')
+  }
+  return otelTracer
+}
+
 function trace ({
   headerSources = ['x-honeycomb-trace', 'x-request-id'],
 } = {}) {
-  if (!process.env.HONEYCOMB_WRITEKEY) {
+  if (!isHoneycomb(process.env)) {
     return (next: Handler) => (context: Context) => next(context)
   }
 
-  const schema = require('honeycomb-beeline/lib/schema')
-  const tracker = require('honeycomb-beeline/lib/async_tracker')
-
-  return function honeycombTrace (next: Handler) {
-    return (context: Context) => {
-      const traceContext = _getTraceContext(context)
-      const trace = beeline.startTrace({
-        [schema.EVENT_TYPE]: 'boltzmann',
-        [schema.PACKAGE_VERSION]: '1.0.0',
-        [schema.TRACE_SPAN_NAME]: `${context.method} ${context.url.pathname}${context.url.search}`,
-        [schema.TRACE_ID_SOURCE]: traceContext.source,
-        'request.host': context.host,
-        'request.original_url': context.url.href,
-        'request.remote_addr': context.remote,
-        'request.method': context.method,
-        'request.scheme': context.url.protocol,
-        'request.path': context.url.pathname,
-        'request.query': context.url.search
-      },
-      traceContext.traceId,
-      traceContext.parentSpanId,
-      traceContext.dataset)
-
-      if (isDev()) {
-        context._honeycombTrace = trace
-      }
-
-      if (traceContext.customContext) {
-        beeline.addContext(traceContext.customContext)
-      }
+  if (isOtel(process.env)) {
+    return function honeycombOtelTrace (next: Handler) {
+      return (context: Context) => {
+        let activeContext = otelContext.active()
```

Comment:
I could be getting confused, but my *understanding* is that Contexts are immutable, meaning if you want to modify the active context you need to grab it, then create a new one with the `propagation.extract` API, then set it with the new value. But again, it's entirely possible I'm missing something here, or picked up a bad pattern from the CI repo.

Reply:
I guess the example you linked with the `W3CTraceContextPropagator` doesn't set the active context, merely passing it to `startSpan`, so maybe I don't need to do the set explicitly. I'll fiddle with it.

Reply:
I switched this step to use context's `with` API. I believe this is appropriate, since we do more with the context than what the propagator would do on its own - I think? Either way the pattern's different now and it's probably time to re-evaluate.